src/test/java/oripa/domain/cptool/LineAdderTest.java

```diff
@@ -34,40 +37,131 @@
  */
 class LineAdderTest {
 
+	public static final OriLine DIAGONAL_LINE = new OriLine(-200, -200, 200, 200, OriLine.Type.MOUNTAIN);
+	private LineAdder adder;
+
 	/**
 	 * Test method for
 	 * {@link oripa.domain.cptool.LineAdder#addLine(oripa.value.OriLine, java.util.Collection)}.
 	 */
 	@Test
 	void testAddLine() {
-		var creasePattern = new ArrayList<OriLine>(List.of(
+		var creasePattern = new ArrayList<>(List.of(
 				new OriLine(0, 0, 100, 0, OriLine.Type.MOUNTAIN),
 				new OriLine(0, 50, 100, 50, OriLine.Type.AUX)));
 
 		var line = new OriLine(20, 50, 20, -10, OriLine.Type.VALLEY);
 
-		var adder = new LineAdder();
-
 		adder.addLine(line, creasePattern);
 
 		assertEquals(6, creasePattern.size());
 	}
 
+	@BeforeEach
+	void setUp() {
+		adder = new LineAdder();
+	}
+
+	@Test
+	void should_split_existing_lines_if_new_line_crosses_them() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(DIAGONAL_LINE));
+		var line = new OriLine(200, 0, -200, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(4, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 200, 200, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, -200, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_remove_duplicated_segments_if_two_lines_of_same_type_are_overlapping() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(new OriLine(0, 0, 200, 0, OriLine.Type.VALLEY)));
+		var line = new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(2, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(100, 0, 200, 0, OriLine.Type.VALLEY)));
```

Comment:
According to the result of tests, this line fails. `Collection::contains` might not be appropriate for the test.

Reply:
lifesbest23@66effa7321b07a0ada28ce818f833ec80453bb3a should resolve this. Regarding your question, I have added something below.